module/state_synchronization/adder.go

```diff
@@ -0,0 +1,285 @@
+package state_synchronization
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/blobs"
+	"github.com/onflow/flow-go/network"
+)
+
+const defaultBlobBatchSize = 16
+
+type ExecutionDataAdderOption func(*executionDataAdderImpl)
+
+func WithBlobSizeLimit(size int) ExecutionDataAdderOption {
+	return func(s *executionDataAdderImpl) {
+		s.maxBlobSize = size
+	}
+}
+
+func WithBlobBatchSize(batchSize int) ExecutionDataAdderOption {
+	return func(s *executionDataAdderImpl) {
+		s.blobBatchSize = batchSize
+	}
+}
+
+func NewExecutionDataAdder(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+	logger zerolog.Logger,
+	metrics module.ExecutionDataAdderMetrics,
+	opts ...ExecutionDataAdderOption,
+) *executionDataAdderImpl {
+	eda := &executionDataAdderImpl{
+		logger.With().Str("component", "execution_data_adder").Logger(),
+		metrics,
+		DefaultMaxBlobSize,
+		defaultBlobBatchSize,
+		&serializer{codec, compressor},
+		blobService,
+	}
+
+	for _, opt := range opts {
+		opt(eda)
+	}
+
+	return eda
+}
+
+var _ ExecutionDataAdder = (*executionDataAdderImpl)(nil)
+
+type executionDataAdderImpl struct {
+	logger        zerolog.Logger
+	metrics       module.ExecutionDataAdderMetrics
+	maxBlobSize   int
+	blobBatchSize int
+	serializer    *serializer
+	blobService   network.BlobService
+}
+
+func (s *executionDataAdderImpl) AddExecutionDataRoot(
+	ctx context.Context,
+	blockID flow.Identifier,
+	chunkExecutionDataCIDs []cid.Cid,
```

Comment:
Taking `chunkExecutionDataCIDs` would assume the caller has called the BlobService to break the chunks execution data into CIDs, but I think this should be part of the Adder's responsibility in order for Adder to guarantee data integrity. Otherwise, the caller might run into mistake that called with wrong Cid slice, which would be hard to debug. 

Refer to my suggested API design in the above comment, where I suggest the Adder should only provide a `AddChunkifiedExecutionData` which takes the ChunkExecutionData, and convert it into slice of CIDs internally.

Reply:
Originally I was gonna do this, but I think I actually want to incorporate the adding of the individual Chunk Execution Data as part of the block execution pipeline, here: https://github.com/onflow/flow-go/blob/7234bce41a91a71fad6244bcd3d70f5491cc905b/engine/execution/computation/computer/computer.go#L149-L185

What do you think about this?

Reply:
I see. I just have one concerns for that:
Adding the execution data is lower priority compare to finish executing the block. If we integrate it into the block execution pipeline, it might slow down the generation of the execution receipt, which will slow down time to seal. Better let EN focus on block execution, and running on a different thread for adding execution data. But there is a challenge here, the progress of adding execution data might be slower than block execution. If we've executed block 1000, but only added block 200, and then crashed/restarted, how to resume adding the executed blocks? how to ensure that eventually all executed block will be added to state sync service?

Reply:
But the point is that the Execution Data ID is part of the Execution Result, so it cannot generate the receipt until it generates this ID anyways

Reply:
I see. But I think only the generation of CIDs is needed for generation the ExecutionDataID, the upload of the execution data should not block the block execution right?

Reply:
Perhaps one way to solve the problem is to redesign the Adder implementation, so that the actual storing of data to the BlobService is done by separate worker routines? So we could just generate and return the CIDs immediately, while the actual data will be stored asynchronously.

Reply:
>but I think I actually want to incorporate the adding of the individual Chunk Execution Data as part of the block execution pipeline, here:

The execution node executes incorporated blocks (not finalized yet). But access nodes usually only care about the execution state for the sealed blocks. When answering user's state query, they are ran against the state from sealed blocks. So it's not necessary to store or sync execution data for blocks that haven't sealed yet. If we integrate adding individual chunk execution data to the block execution pipeline, then we are possibly storing for blocks that eventually been abandoned as they are conflicting with finalized forks. 

So I think we don't have to integrate it into the block execution pipeline. Instead, having another thread going through each finalized block, which eventually will be sealed, and store execution data for them. Thoughts?

Reply:
Another thought: If we are not integrating them, then we probably don't need  `AddChunkExecutionData` to add for each individual chunk, but just adding for all chunks.